listenbrainz/domain/lastfm.py

```diff
@@ -0,0 +1,119 @@
+import uuid
+
+import requests
+from flask import current_app
+from psycopg2.extras import execute_values
+from requests.adapters import HTTPAdapter, Retry
+from sqlalchemy import text
+
+from listenbrainz import db
+from brainzutils import musicbrainz_db
+
+from listenbrainz.webserver.errors import APINotFound
+
+
+def bulk_insert_loved_tracks(user_id: int, feedback: list[tuple[int, str]]):
+    # DO NOTHING because we can only import loved tracks.
+    # if a user has imported something before from last.fm then the score doesn't change.
+    # if the user has unloved a track on last.fm, we won't receive it this time so can't
+    # remove it from feedback in any case.
+    query = """
+        INSERT INTO recording_feedback (user_id, created, recording_mbid, score)
+             VALUES %s
+        ON CONFLICT (user_id, recording_mbid)
+         DO NOTHING
+    """
+    connection = db.engine.raw_connection()
+    with connection.cursor() as cursor:
+        execute_values(cursor, query, feedback, template=f"({user_id}, to_timestamp(%s), %s, 1)")
+    connection.commit()
+
+
+def load_recordings_from_tracks(track_mbids: list) -> dict[str, str]:
+    query = """
+        SELECT track.gid::text AS track_mbid
+             , recording.gid::text AS recording_mbid
+          FROM track
+          JOIN recording
+            ON track.recording = recording.id
+         WHERE track.gid IN :tracks
+    """
+    with musicbrainz_db.engine.connect() as connection:
+        result = connection.execute(text(query), {"tracks": tuple(track_mbids)})
+        return {row["track_mbid"]: row["recording_mbid"] for row in result.mappings()}
+
+
+def fetch_lfm_feedback(lfm_user):
```

Comment:
docstring mssing.

Also, there is potential for a bad user experience here. While it is unlikely that a lot of users have pages upon pages of feedback, this function could take seconds, possibly minutes (in case of retries) to complete, in which the user's browser will time out the page load, leaving the user in the dark about the import of their feedback.

Sadly, there is no simple solution for fixing this. :(

Reply:
Makes sense, discussed on IRC to explore running this import in background in future.